proof/courier.go

```diff
@@ -541,11 +541,17 @@ func (e *BackoffExecError) Error() string {
 
 // BackoffCfg configures the behaviour of the proof delivery backoff procedure.
 type BackoffCfg struct {
-	// SkipInitDelay is a flag that indicates whether we should skip
-	// the initial delay before attempting to deliver the proof to the
-	// receiver.
+	// SkipInitDelay is a flag that indicates whether we should skip the
+	// initial delay before attempting to deliver the proof to the receiver
+	// or receiving from the sender.
 	SkipInitDelay bool
 
+	// InitialReceiveDelay, if not zero, is the amount of time we'll wait
+	// before attempting to receive a proof from the sender. If this is set,
+	// we'll wait for the specified duration regardless of what the transfer
+	// log contains.
+	InitialReceiveDelay time.Duration
```

Comment:
In my mind this delay is more appropriately part of the custodian. Because it relates to the method that the custodian uses to respond to on-chain transactions which indicate an asset transfer has taken place. That is to say, when should the custodian retrieve a proof having seen an on-chain transaction.

This delay isn't related to backoffs and repeat attempts (i.e. `BackoffCfg`).

How do you think about this delay?

Reply:
Yes, you are correct that waiting for the sender to upload the proof using the courier is more of a function of the custodian. But imagine for some reason the delay is super short or super long depending on the type of proof courier, would we then go ahead and add a per-courier delay to the custodian?
I think it would be more easy to understand and configure if all time/backoff related configuration items that actually could be different per proof courier are configured on the courier itself.
But happy to change this back if you disagree.

Reply:
I agree with Oli here, mostly because I think these timeouts should not be duplicated across multiple packages / subsystems, and since the courier configs already define timeout & backoff behavior, IMO it works better to have those values in those configs.

Another way to think about it would be, if I'm adding a courier with timeouts and backoff delays very different to what we have now, what would I have to change & where? IIUC this PR moves all of those potential changes into the courier config and internal logic.

Reply:
I'm not sure.

These changes hard code the delay into every new instance of the custodian's proof courier types. But if the custodian is [handling an old transaction](https://github.com/lightninglabs/taproot-assets/blob/becddbf4a658ef3523ebbb4570daaaac1066b944/tapgarden/custodian.go#L286) (catching up after being offline) then the proof courier will unnecessarily delay proof retrieval. I think this highlights my point that waiting on the timeout is a custodian specific feature.

Sure we could add precision by making the delay proof courier specific. But whether we want the delay or not is based on when we start handling a receive event.

For example, we might use the custodian to retrieve proofs as part of recovery. In which case we wouldn't want this delay at all.

I think the custodian needs to decide on whether to delay or not and the proof courier needs to decide on how long that delay is.

And I'm still not convinced this should be part of the backoff config. Right now, as far as I can tell, no part of the backoff procedure is sender/receiver specific. I think it would be good to retain and enhance that generality if possible. This delay could happen before starting the backoff procedure.

Reply:
Okay, I guess that's a solid argument. I removed the change, please take a look.